---
title: "Ratkevich_proteomics_HW"
author: "Ratkevich Ilya"
date: "2024-04-18"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include=FALSE}
library(readxl)
library(limma)
library(ape)
library(dendextend)
library(RColorBrewer)
library(gplots)
library(NMF)
library(vegan)
library(Biobase)
library(DT)
library(ggplot2)
library(gprofiler2)
library(impute)
library(ggrepel)
library(clusterProfiler)
```

### Upload the data

```{r include=FALSE}
peaks_data <- read.csv('/home/ilya/Downloads/peaks_data.csv')
```

## EDA

Remove genes with half and more missing values

```{r}
peaks_data <- peaks_data[which(rowMeans(!is.na(peaks_data)) > 0.5), ]
peaks_data <- peaks_data[which(!duplicated(peaks_data$Gene_id)),]
rownames(peaks_data) <- peaks_data$Gene_id
peaks_data <- peaks_data[,-c(1, 2, 3)]
```

Imputation of all the missing values:

```{r}
peaks_data <- as.matrix(peaks_data)
peaks_data_trans <- t(peaks_data)
knn_peaks <- impute.knn(peaks_data_trans, k = 2)
knn_peaks <- knn_peaks$data
knn_peaks <- as.data.frame(knn_peaks)
knn_peaks <- t(as.matrix(knn_peaks))

#yeast_experiment <- as.factor(c("wt_1", "wt_1", "wt_1", "wt_1", "105_2", "105_2", "105_2", "105_2", "wt_2", "wt_2", "105_3", "105_3", "105_3", "105_3", "wt_3", "wt_3", "wt_3", "wt_3"))

peaks_experiment <- as.factor(c("1", "1", "1", "1", "1","1","1","1","1","1","1","1","1","1", "2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2","2"))
```

```{r}
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[peaks_experiment]
boxplot(knn_peaks, outline = FALSE, main = "Исходные данные", col = cols)
legend("topright", levels(peaks_experiment), fill = pal, bty = "n", xpd = T)
```

Делаем нормализацию данных -- логарифмирование + квантильная нормализация

```{r}
peaks_log <- log2(knn_peaks)
peaks_norm <- normalizeQuantiles(as.matrix(peaks_log))
boxplot(peaks_norm, outline = FALSE, main = "Нормализованные данные", col = cols)
legend("topright", levels(peaks_experiment), fill = pal, bty = "n", xpd = T)
```

```{r}
peaks_pca <- t(peaks_norm)
terapod_pca <- rda(peaks_pca, scale = TRUE)

rownames(peaks_pca) <- colnames(knn_peaks)

df_scores <- data.frame(peaks_pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))

p_scores <- ggplot(df_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(peaks_pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3)) + ggtitle(label = "Ordination") + theme_bw()
p_scores
```

MA-plot для сравения групп

```{r}
maplot <- function(X1, X2, pch = 21, main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", lpars = list(col = "blue", lwd = 2), ...){
  # Координаты
  X <- (rowMeans(X2) + rowMeans(X1)) / 2
  Y <- rowMeans(X2) - rowMeans(X1)
  # График
  scatter.smooth(x = X, y = Y, main = main, pch = pch, xlab = xlab, ylab = ylab, lpars = lpars, ...)
  abline(h = c(-1, 0, 1), lty = c(2, 1, 2))
}

```


```{r}
peaks_norm_clean <- data.frame(peaks_data)
peaks_norm_clean[sapply(peaks_norm_clean, is.infinite)] <- NA
peaks_norm_clean <- na.omit(peaks_norm_clean)
maplot(peaks_norm_clean[,1:14], peaks_norm_clean[,15:23])
```

### Differential expression 

```{r}
# Differential expression 
expr_data <- as.matrix(peaks_norm)

# Samples data
pheno_data <- data.frame(peaks_experiment)
rownames(pheno_data) <- colnames(peaks_norm)
pheno_metadata <- data.frame(
  labelDescription = c("Experimental condition"),
  row.names=c("Condition"))
pheno_data <- new("AnnotatedDataFrame",
                 data = pheno_data,
                 varMetadata = pheno_metadata)

# Protein data
feature_data <- data.frame(Prot = rownames(expr_data))
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
  labelDescription = c("Protain name"),
  row.names = c("Protain"))
f_data <- new("AnnotatedDataFrame",
              data = feature_data,
              varMetadata = feature_metadata)

# Combine together
exp_set <-
  ExpressionSet(assayData = expr_data,
                phenoData = pheno_data,
                featureData = f_data)
```


```{r}
X <- model.matrix(~ peaks_experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)
efit <- eBayes(fit)
```


```{r}
MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...){
  # соотношение и интенсивность
  R <- efit$coefficients[, coef]
  I <- efit$Amean
  # прозрачный цвет
  col_btransp <- adjustcolor(col, alpha.f = alpha)
  # график
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  # отмечаем дифференциально-экспрессируемые белки
  if(signif){
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex*2, col = "orange2", pch = pch.signif)
  }
  # подписываем первые n белков с сильнее всего различающейся экспрессией
  if(text){
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}


MA_limma(efit, coef = 2, n = 30)
```


```{r}
# ## Добываем дифференциально-экспрессируемые белки для дальнейшей работы =======
# Первые 20 дифференциальных белков
my_list <- topTable(efit, coef = 2, n = 100)
# Фильтруем ExpressionSet
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```


```{r fig.height=10, fig.width=15}
dat <- as.matrix(exprs(dif_exp_set))
dat[sapply(dat, is.infinite)] <- NA
dat <- na.omit(dat)
# после дополнительной стандартизации по белкам
pal_blue_red <- colorpanel(75, low = "steelblue", mid = "black", high = "red")
heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(3, 0.1, 3, 0.1)))
```

```{r}
topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
full_list <- topTable(efit, number = numGenes)
full_list <- full_list[full_list$adj.P.Val <= 0.05,]
#write.csv(full_list, '/home/Ilya/Downloads/DE_full')
```

```{r}
MA_limma(efit, coef = 2, n = 4)
```


### MA-plot первых 20 дифференциально экспрессируемых белков, но таких, чтобы уровень экспрессии различался в 2 раза

```{r}
# MA-plot первых 20 дифференциально экспрессируемых белков, но таких, чтобы уровень экспрессии различался в 2 раза
MA_limma(efit, coef = 2, n = 80, text = F, lfc = 1)
```



```{r}
# ## Добываем дифференциально-экспрессируемые белки для дальнейшей работы =======
# Первые 20 дифференциальных белков
my_list <- full_list
# Фильтруем ExpressionSet
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```

# Volcano plot для дикого типа и дикого типа 1-ой стадии

```{r }
volcano_list <- full_list

volcano1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()

volcano2 <- volcano1 + geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$diffexpressed <- "NO"       
```

сортировка данных, если logFC > 1 - up-регуляция, по идее если логФС равен 1, то у нас экспрессия в два раза изменилась, ну и значимость делаем адекватную (можно 0.08)

```{r }
volcano_list$diffexpressed[volcano_list$logFC > 1 & volcano_list$adj.P.Val < 0.05] <- "UP"
```

 if log2Foldchange < -1 and pvalue < 0.05, set as "DOWN" (можно -0.06)

```{r }
volcano_list$diffexpressed[volcano_list$logFC < -1 & volcano_list$adj.P.Val< 0.05] <- "DOWN"

volcanodif1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed)) + geom_point() + theme_minimal()
```

 Add lines as before...

```{r }
volcanodif2 <- volcanodif1 + geom_vline(xintercept=c(-1, 1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$Prot <- as.character(volcano_list$Prot)
volcano_list$delabel <- NA
volcano_list$delabel[volcano_list$diffexpressed != "NO"] <- volcano_list$Prot[volcano_list$diffexpressed != "NO"]
#volcano_list[volcano_list$adj.P.Val<=0.6,]$delabel <- volcano_list[volcano_list$adj.P.Val<=0.6,]$Prot

plot_proteins <- ggplot(data=volcano_list, aes(x=logFC, y=-log10(adj.P.Val), col=diffexpressed, label=delabel)) + 
  geom_point() + 
  theme_minimal() +
  geom_text_repel(size=3, colour = 'black', max.overlaps = 30)

plot_final <- plot_proteins + geom_vline(xintercept=c(-1, 1), col="red") + geom_hline(yintercept=-log10(0.05), col="red")
plot_final
```

## Дерево и ординация по дифэкспрессируемым белкам

```{r fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
diff_prot <- rownames(full_list)
diff_expressed <- as.data.frame(peaks_norm)[diff_prot,]
t_diff_expressed <- t(diff_expressed)
#rownames(t_diff_expressed) <-  as.factor(gsub("_[^_]*", replacement = "", rownames(t_diff_expressed)))
#rownames(t_diff_expressed) <- make.unique(as.character(pheno$yeast_experiment))

d <- dist(x = t_diff_expressed, method = "canberra")

mouse_hc_avg <- hclust(d, method = "average")
mouse_ph_avg <- as.phylo(mouse_hc_avg)
mouse_den_avg <- as.dendrogram(mouse_hc_avg)

get_colours <- function(dend, n_chars, palette = "Dark2"){ #nchars = первые нескольок симовлов которые будут использоваться для разделения фактора на группы
labs <- get_leaves_attr(dend, "label")
group <- substr(labs, start = 0, stop = n_chars)
group <- factor(group)
cols <- brewer.pal(length(levels(group)), name = palette)[group]
return(cols)
}

cols <- get_colours(dend = mouse_den_avg, n_chars = 6)
den_avg_c <- color_labels(dend = mouse_den_avg, col = cols)
plot(den_avg_c, horiz = TRUE)
```


# GO enrichment analysis
In order to group significantly differentially expressed proteins we have used GO enrichment analysis.

```{r warning=FALSE, message=FALSE}
# keep only the significant proteins results
sig <- subset(full_list, adj.P.Val < 0.05)
# get the significant up-regulated proteins
up <- subset(sig, logFC > 0)
# get the significant down-regulated proteins
down <- subset(sig, logFC < 0)
```

```{r warning=FALSE, message=FALSE}
# needed to convert to enrichResult object
up_names <- gconvert(row.names(up))
down_names <- gconvert(row.names(down))
```

## Up-regulated proteins
```{r warning=FALSE, message=FALSE}
# enrichment analysis using proteins names
multi_gp_up_reg <- gost(list("up-regulated" = up_names$name), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_up = multi_gp_up_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_up <- gp_mod_up[order(gp_mod_up$p_value), ]
gp_mod_up_BP <- gp_mod_up[gp_mod_up$source == "GO:BP", ]
gp_mod_up_BP$GeneRatio <- paste0(gp_mod_up_BP$intersection_size,  "/", gp_mod_up_BP$query_size)
gp_mod_up_BP$BgRatio <- paste0(gp_mod_up_BP$term_size, "/", gp_mod_up_BP$effective_domain_size)
names(gp_mod_up_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_up_BP$geneID <- gsub(",", "/", gp_mod_up_BP$geneID)
#row.names(gp_mod_up_BP) <- gp_mod_up_BP$ID
gp_mod_enrich_up_BP <- new("enrichResult", result = gp_mod_up_BP)
```

Draw enrichment plot:
```{r warning=FALSE, message=FALSE}
enrichplot::dotplot(gp_mod_enrich_up_BP, showCategory = 10) + ggplot2::labs(title = "up-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")
```


## Down-regulated proteins
```{r warning=FALSE, message=FALSE}
# enrichment analysis using gene names
multi_gp_down_reg <- gost(list("down-regulated" = down_names$name), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_down = multi_gp_down_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_down <- gp_mod_down[order(gp_mod_down$p_value), ]
# BP
gp_mod_down_BP <- gp_mod_down[gp_mod_down$source == "GO:BP", ]
gp_mod_down_BP$GeneRatio <- paste0(gp_mod_down_BP$intersection_size,  "/", gp_mod_down_BP$query_size)
gp_mod_down_BP$BgRatio <-  paste0(gp_mod_down_BP$term_size, "/", gp_mod_down_BP$effective_domain_size)
names(gp_mod_down_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_down_BP$geneID <- gsub(",", "/", gp_mod_down_BP$geneID)
gp_mod_enrich_down_BP <- new("enrichResult", result = gp_mod_down_BP)
```

Draw enrichment plot:
```{r warning=FALSE, message=FALSE}
enrichplot::dotplot(gp_mod_enrich_down_BP, showCategory = 20) + ggplot2::labs(title = "down-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")
```

The pathogenesis of AVC involves a multifactorial process that includes:

 Endothelial dysfunction: Damage to the endothelial cells lining the valve triggers an inflammatory response that leads to the recruitment of macrophages and other immune cells. Oxidative stress: Oxidative stress from free radicals and reactive oxygen species can damage valve cells and promote inflammation and calcification. 
 There seem to be some stress in cells, cince some processes arfe activated, like generation of precursors, protein translation, different transport
 Dysregulation of calcium metabolism: Abnormal regulation of calcium transport and deposition by valve cells contributes to the accumulation of calcium deposits. There seem no trail of these paths
 Osteogenic transdifferentiation: Valve cells can undergo a phenotypic switch to become osteoblast-like cells, promoting the formation of bone-like tissue on the valve. It's a part of cell organisation processes
 Inflammation and immune response: Chronic inflammation and the recruitment of inflammatory cells play a significant role in valve calcification by releasing pro-calcific cytokines and chemokines. Compound transport and other transport types include these pathways
